File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.kt
import kotlin.math.sqrt

class Solution {
    val permutations = mutableListOf<List<Int>>()

    fun solution(numbers: String): Int {
        setPermutation(numbers.map { it - '1' + 1 })

        return permutations.drop(1)
            .map { it.joinToString("").toInt() }
            .distinct()
            .filter { isPrime(it) }
            .size
    }

    fun setPermutation(input: List<Int>, final: List<Int> = listOf(), sub: List<Int> = input): List<List<Int>> {
        permutations.add(final)

        return if (sub.isEmpty()) {
            listOf(final)
        } else {
            sub.flatMap { setPermutation(input, final + it, sub - it) }
        }
    }
    
    fun isPrime(input: Int): Boolean {
        return if (input <= 1) {
            false
        } else {
            (2..sqrt(input.toDouble()).toInt()).none { input % it == 0 }
        }
    }
}

File: 프로그래머스/lv2/42839. 소수 찾기/소수 찾기.py
from itertools import permutations

def solution(numbers):
    per = []
    
    for i in range(1, len(list(numbers)) + 1):
        per += list(permutations(numbers, i))
        
    nums = set([int("".join(p)) for p in per])
    cnt = 0
    
    for i in nums:
        check = True
        
        if i < 2:
            continue
        
        for j in range(2, int(i ** 0.5) + 1):
            if i % j == 0:
                check = False
                break
    
        if check:
            cnt += 1
    
    return cnt